src/rest_framework_dso/embedding.py

```diff
@@ -77,6 +83,96 @@ def _expand_parse_error(allowed_names, field_name):
+    is preferred here as it returns instances while reading them. Nothing is stored in memory.
+    Hence, both approaches are fundamentally incompatible. This class performs a
+    mixed strategy: load a chunk, and perform prefetches for that particular batch.
+
+    As extra performance benefit, a local cache avoids prefetching the same records
+    again when the next chunk is analysed. It has a "least recently used" cache to avoid
+    flooding the caches when foreign keys constantly point to different unique objects.
+    """
+
+    def __init__(self, queryset: models.QuerySet, chunk_size=None, sql_chunk_size=None):
+        """
+        :param queryset: The queryset to iterate over, that has ``prefetch_related()`` data.
+        :param chunk_size: The size of each segment to analyse in-memory for related objects.
+        :param sql_chunk_size: The size of each segment to fetch from the database,
+            used when server-side cursors are not available. The default follows Django behavior.
+        """
+        self.queryset = queryset
+        self.sql_chunk_size = sql_chunk_size or DEFAULT_SQL_CHUNK_SIZE
+        self.chunk_size = chunk_size or self.sql_chunk_size
+        self._fk_caches = defaultdict(lambda: LRU(self.chunk_size // 2))
+
+    def __iter__(self):
+        # Using iter() ensures the ModelIterable is resumed with the next chunk.
+        qs_iter = iter(self.queryset.iterator(chunk_size=self.sql_chunk_size))
+
+        # Keep fetching chunks
+        while instances := list(islice(qs_iter, self.chunk_size)):
+            # Perform prefetches on this chunk:
+            if self.queryset._prefetch_related_lookups:
+                self._add_prefetches(instances)
+            yield from instances
+
+    def _add_prefetches(self, instances: List[M]):
+        """Merge the prefetched objects for this batch with the model instances."""
+        if self._fk_caches:
+            # Make sure prefetch_related_objects() doesn't have to fetch items again
+            # that infrequently changes (e.g. a "wijk" or "stadsdeel").
+            all_restored = self._restore_caches(instances)
+            if all_restored:
+                return
+
+        # Reuse the Django machinery for retrieving missing sub objects.
+        # and analyse the ForeignKey caches to allow faster prefetches next time
+        models.prefetch_related_objects(instances, *self.queryset._prefetch_related_lookups)
+        self._persist_prefetch_cache(instances)
+
+    def _persist_prefetch_cache(self, instances):
+        """Store the prefetched data so it can be applied to the next batch"""
+        for instance in instances:
+            for lookup, obj in instance._state.fields_cache.items():
+                if obj is not None:
+                    cache = self._fk_caches[lookup]
+                    if obj.pk not in cache:
+                        cache[obj.pk] = obj
```

Comment:
Is there a reason why you're looking before you leap? This way, the cache is no longer truly LRU. E.g.,

```python
d = LRU(2)
for key in "foo bar baz foo quux foo bar".split():
    if key not in d:
        d[key] = None
```

stores `["bar", "quux"]` in the cache, rather than `["bar", "foo"]`.

Reply:
Thanks, fixed!